### ISSUE TITLE
fix (or ameliorate) changed RHS handler contract

### ISSUE THREAD

**Issue by garfieldnate** [SoarGroup/Soar#369]
We merged this PR in order to provide a cross-platform C ABI for adding RHS functions: https://github.com/SoarGroup/Soar/pull/341. However, in the process we totally broke all existing C++ RHS functions. One group has 200+, and would like a solution simpler than just "add this code to the beginning and end of all your RHS handlers".

I hoped this would be as simple as adding a new `register` function that simply wraps handlers that return `std::string` and returns a handler that works with the `char* buf` parameter, but this is not possible in C++ because you can't convert lambda functions that capture variables (we capture the handle we are wrapping) into function pointers. Here's my non-functional implementation here: https://github.com/SoarGroup/Soar/commit/b0687473bec348c99f2ab7f45af8fb0e915bc8d9.

My next thought is to add a flag to RHS handlers (the struct is accessed before calling [here](https://github.com/SoarGroup/Soar/blob/9530289cc5452fc3a4c75cf5b5f4ca25cb46fde4/Core/ClientSML/src/sml_ClientKernel.cpp#L698)). The flag would indicate to Soar that the function uses either the C or C++ API. I dislike the increase in complexity with this solution.

I'm partly inclined to just revert the API changes made in 9.6.1, fixing all clients except, unfortunately, SoarTech, which would have to continue maintaining its own patches.

Any ideas here @marinier?

**Comment by marinier** [comment]
Can you use a std::Function instead of a lambda as described here?

https://stackoverflow.com/questions/28746744/passing-capturing-lambda-as-function-pointer/28747100#28747100

There is some slight overhead in doing this, but I'd think this is ok as it's just a workaround for the better way of doing it anyway.

Otherwise I'd vote for adding the flag, I guess, although it would be nice if there was some way to emit a compiler warning when people used it to indicate that this approach is deprecated.

If all else fails, I'd suggest meeting with the affected party to discuss converting their functions, because it should be really easy so I don't understand what the fuss is about. Even for 200 functions they could be done in less than a day.

**Comment by garfieldnate** [comment]
The `std::function` solution would replace the function pointers, not the lambda, meaning that the handlers themselves would need to be defined as `std::function`s, making it a non-C API again, and also requiring changes by all parties.

The fuss... well I hadn't known that there were some who extensively use the C++ API, and we did make a breaking change on a minor version release. The C API is less clear now, and the reason for it is sort of unclear since there are still other SML handler methods that are specified to return `std::string`. The C API requires that you carefully read the documentation to use it correctly, unlike the C++ API, for which you could simply read the signature types, so it feels like a strange usability downgrade.

I suppose another possibility to look at would be macros. We could definitely provide macros for the recommended before/after code blocks. Would be even nicer if we could use a macro to wrap the C++ handler functions and to create C versions.

**Comment by marinier** [comment]
What's strange to me is that Soar needs to be compiled for every Linux runtime library that someone wants to use. That's fine for a university project but not for real world usage. This change is a first step in hopefully defining a C API for the entire SML so that this is possible. When SML was first written, this was the only place that wasn't a pure C API; it's unfortunate that this problem has spread.

I agree that changing things on a micro version number isn't great, but Soar has never defined its version semantics. Maybe it's time to adopt something more formal, like [semantic versioning](https://semver.org/).

Another approach to supporting both the C and C++ APIs could be having two separate sets of RHS functions, the C ones and the C++ ones. Not great, but it avoids the problem of trying to translate one to the other.

This is just my 2 cents; obviously I'm not a decision maker here. To me the decision boils down to whether Soar is on a path to becoming a system appropriate for real world usage or whether it is going to mostly stay a student-grown project. These sorts of engineering concerns, like providing a C API and version semantics, are a necessary part of "growing up" the system. But it may be that this isn't John's goal, which is fine. (I'd personally be disappointed, but it's not the end of the world either.)

**Comment by scijones** [comment]
Essentially an outsider to some of these concerns, but from my simplistic point of view:

> "Another approach to supporting both the C and C++ APIs could be having two separate sets of RHS functions, the C ones and the C++ ones. Not great, but it avoids the problem of trying to translate one to the other."

Sounds great, despite the obvious inelegance. Backwards compatibility in some form for current ongoing users who are upset about rewriting functions, forwards compatibility in the way expected for "real-world" cases. And going forwards we'd treat the old way as legacy support (wouldn't do anything more with it) and just focus on the new way.

**Comment by garfieldnate** [comment]
Okay, this isn't nearly as complicated as I thought it would be. 

>But it may be that this isn't John's goal

This is all me making it more complicated than it needs to be. No fault of John's XD.

I pushed a branch, `rhs_handler_plan_B`, that isn't finished but I think the idea is there:

* Leave the new RHS handler C API in place. This is an asset when linking against dynamic libraries, which is how Soar is used with other languages.
* While leaving the C-API in place, use C++ std::function internally in Soar.
* Add *new* APIs with the same names but ending in `CPP` (I'm open to different names). These API's will simply accept `std::function` handlers.

So the [event handler definitions](https://github.com/SoarGroup/Soar/blob/rhs_handler_plan_B/Core/ClientSML/src/sml_ClientEvents.h#L105) are like so:

```C++
    typedef char const *(*RhsEventHandler)(smlRhsEventId id, void* pUserData, Agent* pAgent,
                                          char const* pFunctionName, char const* pArgument, int *buffSize, char *buff) ;

    using RhsEventHandlerCPP = std::function<std::string(smlRhsEventId id, Agent *pAgent, char const *pFunctionName, char const *pArgument)>;
```

And then in [`Kernel::AddRhsFunction`](https://github.com/SoarGroup/Soar/blob/rhs_handler_plan_B/Core/ClientSML/src/sml_ClientKernel.cpp#L2137), we convert the function pointer and pUserData pointer into a `std::function` with the userData captured and the char* re-allocation handled:

```C++
RhsEventHandlerCPP c2cppHandler(RhsEventHandler handler, void* pUserData) {
    return [handler, pUserData](
        smlRhsEventId id,
        Agent* pAgent,
        char const* pFunctionName,
        char const* pArgument) -> std::string {
        char buffer[SML_INITIAL_BUFFER_SIZE] = {0};
        char *buff = buffer;
        std::string result;
        int buffSize = sizeof(buffer);

        const char *retVal = handler(id, pUserData, pAgent, pFunctionName, pArgument, &buffSize, buff) ;

        if ( !retVal )
        {
            buff = new char[buffSize];

            result = handler(id, pUserData, pAgent, pFunctionName, pArgument, &buffSize, buff) ;
            // TODO: check for NIL again and print error of some kind
            delete [] buff;
        }
        else {
            result = retVal;
        };
        return result;
    };
}
```

Then the much simpler API for adding handlers from C++ code:

```C++
int Kernel::AddRhsFunctionCPP(char const* pRhsFunctionName, RhsEventHandlerCPP handler, bool addToBack)
```

No need to pass pUserData here, as the client can capture whatever they need in the `std::function` that they provide. This should play better with `std::bind` use-cases as well (requested by @moschmdt).

I think this should be ergonomic for C++ usage while still providing the C API for the folks that need it.

Thank you for the valuable input so far! Any further thoughts/concerns on this are welcome, such as ideas on the naming if *CPP fs not satisfactory. I'm uncertain about whether we might be able to overload the original `AddRhsFunction` without messing up the C API, which would be even nicer and would just leave the naming of `RhsEventHandlerCPP` to think about.

**Comment by marinier** [comment]
Glad to hear there's a good resolution here. Personally I prefer methods
that don't have multiple capital letters in a row (Cpp instead of CPP), but
this is hardly a big deal and it's more important to follow whatever
existing patterns there are (if any).

**Comment by garfieldnate** [comment]
So just an FYI here, as far as I can tell, SWIG generates pure C wrappers for linking in other languages, and we have typemaps for `std::string` loaded, so SWIG is handling the conversion and using `std::string` in an SML API will not cause the final lib to be anything other than pure C.

There's a compiler flag, `-Wextern-c-compat`, which warns if anything not compatible with pure C ABI is used in the SML library. This warning is on by default at least in clang, so we will see them in the build if we accidentally make a non-portable build of an SML lib.

**Comment by garfieldnate** [comment]
I ended up taking a middle ground: keep the nice new API's, revert the change to the old API's. No helper functions will be necessary. Code that broke in 9.6.1 will work again in 9.6.2.

I discussed this extensively with Bob M., but SML is so far removed from a pure C API that we would need a tool like SWIG to wrap it properly. Small changes to the API will only inconvenience users, and will not bring us closer to a portable libSoar by themselves. We can address compiled libSoar portability in a dedicated project if desired.

Changes required in ST projects that used the C-style API will be minimal; they just need to return a `std::string` from their functions, and they can remove all logic related to buffer size and function re-calling.
